model/flow/collection.go

```diff
@@ -67,10 +74,31 @@ func (c Collection) Len() int {
 
 // LightCollection contains cryptographic commitments to the constituent transactions instead of transaction bodies.
 // It is used for indexing transactions by collection and for computing the collection fingerprint.
+//
+//structwrite:immutable - mutations allowed only within the constructor
 type LightCollection struct {
 	Transactions []Identifier
 }
 
+// UntrustedLightCollection is an untrusted input-only representation of a LightCollection,
+// used for construction.
+//
+// This type exists to ensure that constructor functions are invoked explicitly
+// with named fields, which improves clarity and reduces the risk of incorrect field
+// ordering during construction.
+//
+// An instance of UntrustedLightCollection should be validated and converted into
+// a trusted LightCollection using NewLightCollection constructor.
+type UntrustedLightCollection LightCollection
+
+// NewLightCollection constructs a new LightCollection instance.
+// Any errors indicate the input cannot be used to construct a valid LightCollection.
+func NewLightCollection(untrusted UntrustedLightCollection) (*LightCollection, error) {
+	return &LightCollection{
+		Transactions: untrusted.Transactions,
+	}, nil
```

Comment:
Maybe we could add a check to make sure none of the `transaction IDs` are zero? If we go with this, could you please also add a corresponding test case to cover this.

```suggestion
	for i, txID := range untrusted.Transactions {
		if txID == ZeroID {
			return nil, fmt.Errorf("transaction ID at index %d must not be zero", i)
		}
	}
	return &LightCollection{
		Transactions: untrusted.Transactions,
	}, nil
```

Reply:
Thank you for pointing this out. I find Zero IDs sit in a grey area:
- Usually, we never expect a non-optional Identifier field to be Zero (conceptually Zero means "not set"). It is similar to a nil non-optional pointer field.
- However, a zero ID is safer than a nil pointer field, in the sense that it is very unlikely to cause subsequent business logic to crash. 

Broadly speaking, I see two approaches to implementing the constructor validation:
1. **Structural Validity**: Disallow nil pointers, or incorrect indices into arrays, etc (anything that would cause subsequent business logic to crash) ([Example: TimeoutObject nil check](https://github.com/onflow/flow-go/pull/7433/files#diff-7dca66862de2c90c8341d1bf4865b4764887a785f013744962713ee053f434deR82-R84))
2. **Context-Free Validity**: Disallow anything that we know to be an invalid representation of the model which can be determined in-place without additional context ([Example: TimeoutObject checks](https://github.com/onflow/flow-go/pull/7433/files#diff-7dca66862de2c90c8341d1bf4865b4764887a785f013744962713ee053f434deR85-R117))

**Structural Validity** is the main point of having the validation, because much of our subsequent business logic does make assumptions about structural validity. Writing business logic that operates on these models is much easier if we can safely assume that basic structural validity is enforced at an earlier point in the pipeline. **These checks are necessary for constructors.**

**Context-Free Validity** is largely duplicating checks which should already exist in subsequent business logic. These are less valuable but, because the checks are almost always very cheap, they aren't costly either. It's potentially beneficial to throw out known-invalid data earlier in the pipeline, but it's non-ideal to duplicate validation in two places. It's a mixed bag. **These checks are optional for constructors, on a case-by-case basis.**

In this case I feel we gain quite a lot by not doing the context-free validity check here (especially when also taking Yurii's suggestion to eliminate the unused error return), so I'm happy to exclude it.